test/EndToEndTests/Tests/Core/Microsoft.OData.Core.E2E.Tests/WriteJsonPayloadTests/SerializationInfoEdgeTests.cs

```diff
@@ -0,0 +1,552 @@
+        }
+
+        [Theory]
+        [InlineData(MimeTypeODataParameterFullMetadata, true)]
+        [InlineData(MimeTypeODataParameterMinimalMetadata, true)]
+        [InlineData(MimeTypeODataParameterNoMetadata, false)]
+        public async Task SpecifyingSerializationInfoForBothFeedAndEntry_WithMetadata_ShouldUseEntryType(string mimeType, bool containsExpectedTypeInfo)
+        {
+            var settings = CreateODataMessageWriterSettings();
+            var responseMessage = CreateTestResponseMessage(mimeType);
+
+            using (var messageWriter = new ODataMessageWriter(responseMessage, settings))
+            {
+                var odataWriter = await messageWriter.CreateODataResourceSetWriterAsync();
+                var feed = CreatePersonFeed();
+                feed.SetSerializationInfo(
+                    new ODataResourceSerializationInfo()
+                    {
+                        NavigationSourceName = "People",
+                        NavigationSourceEntityTypeName = NameSpacePrefix + "Person"
+                    });
+
+                var entry = new ODataResource()
+                {
+                    Id = new Uri(_baseUri + "People(-5)"),
+                    TypeName = NameSpacePrefix + "Employee"
+                };
+
+                var personEntryP1 = new ODataProperty { Name = "PersonId", Value = -5 };
+                var personEntryP2 = new ODataProperty
+                {
+                    Name = "Name",
+                    Value = "xhsdckkeqzvlnprheujeycqrglfehtdocildrequohlffazfgtvmddyqsaxrojqxrsckohrakdxlrghgmzqnyruzu"
+                };
+
+                var personEntryP3 = new ODataProperty
+                {
+                    Name = "ManagersPersonId",
+                    Value = -465010984
+                };
+
+                entry.Properties = new[] { personEntryP1, personEntryP2, personEntryP3 };
+                entry.SetSerializationInfo(
+                    new ODataResourceSerializationInfo()
+                    {
+                        NavigationSourceName = "Person",
+                        NavigationSourceEntityTypeName = NameSpacePrefix + "Person"
+                    });
+
+
+                await odataWriter.WriteStartAsync(feed);
+                await odataWriter.WriteStartAsync(entry);
+                await odataWriter.WriteEndAsync();
+                await odataWriter.WriteEndAsync();
+            }
+
+            Stream stream = await responseMessage.GetStreamAsync();
+            string result = await TestsHelper.ReadStreamContentAsync(stream);
+
+            if (containsExpectedTypeInfo)
+            {
+                Assert.Contains(NameSpacePrefix + "Employee", result);
+                Assert.Contains("$metadata#People", result);
+            }
+            else
+            {
+                Assert.DoesNotContain(NameSpacePrefix + "Employee", result);
+            }

```

Comment:
This also looks like another case where separate test cases should be considered.